clr/commands.py

```diff
@@ -135,45 +131,149 @@ class Namespace:
 
     @property
     def commands(self):
+        """Sorted list of command names in this namespace."""
         return sorted(self.command_specs.keys())
 
+    def parse_args(self, command_name, argv):
+        """Parse args for the given command."""
+
+        signature = self.command_specs[command_name].signature
+        # Parse the command line arguments, starting after command name.
+        parsed = NoneIgnoringArgparseDestination()
+        self.argument_parser(command_name).parse_args(sys.argv[2:], namespace=parsed)
+
+        # Turn parsed args into somethign we can pass to signature.bind.
+        args = []
+        kwargs = {}
+        for param in signature.parameters.values():
+            value = getattr(parsed, param.name, None)
+            if param.kind  == param.POSITIONAL_ONLY:
+                args.append(value)
+            elif param.kind == param.VAR_POSITIONAL:
+                args.extend(value)
+            elif param.kind in (param.KEYWORD_ONLY, param.POSITIONAL_OR_KEYWORD):
+                if param.default != param.empty and value is None:
+                    # BoundArguments will apply the defaults.
+                    continue
+                kwargs[param.name] = value
+
+        # Ensure the signature is valid and applies default. Could use argparse to do more of this,
+        # but adds correctness guarantees and gives a nice error message when something is wrong.
+        bound_args = signature.bind(*args, **kwargs)
+        bound_args.apply_defaults()
+        return bound_args
+
+    def argument_parser(self, command_name):
+        """Returns an ArgumentParser matching the signature of command.
```

Comment:
There's definitely a lot going on in this method.

I'll confess that I've never really used `argparse` for anything more than building simple shell utils. Could you elaborate why our particular use case requires all these nested conditionals to handle different param types & such?

Reply:
I think this is based on trying to give callers some options, while accepting the function signature's preferences. For example; if there's a function signature like:
```
def cmd_stuff(self, hostname, *, count=5):
```
This function seems like it would turn it into something that can take `hostname` as a positional or named `--hostname` argument, while also requiring `count` to be named-only via: `--count`.

So these would work:
```
clr whatever:stuff hostname --count=1
clr whatever:stuff hostname
clr whatever:stuff  --count=1 hostname
clr whatever:stuff --hostname=hostname --count=1
clr whatever:stuff --hostname=hostname
clr whatever:stuff  --count=1 --hostname=hostname
```

But this should fail (or be nonsense), because `--count` is not allowed to be positional.
```
clr whatever:stuff hostname 1
clr whatever:stuff --hostname=hostname 1
clr whatever:stuff  1 hostname
clr whatever:stuff  1 --hostname=hostname
```

Reply:
But yeah, an example or two would be good. Please feel free to steal my examples if they are correct :)

Reply:
Exactly. Adding more docs.

```
hostname=hostname count=1
+ clr stuff hostname
hostname=hostname count=5
+ clr stuff --count=1 hostname
hostname=hostname count=1
+ clr stuff --hostname=hostname --count=1
hostname=hostname count=1
+ clr stuff --hostname=hostname
hostname=hostname count=5
+ clr stuff --count=1 --hostname=hostname
hostname=hostname count=1
+ echo '# Should start failing'
# Should start failing
+ clr stuff hostname 1
hostname=hostname count=1
+ clr stuff --hostname=hostname 1
usage: clr system:stuff [--hostname HOSTNAME] [--count COUNT]
                        [hostname] [count]
clr system:stuff: error: argument hostname: not allowed with argument --hostname
+ clr stuff 1 hostname
usage: clr system:stuff [--hostname HOSTNAME] [--count COUNT]
                        [hostname] [count]
clr system:stuff: error: argument count: invalid int value: 'hostname'
+ clr stuff 1 --hostname=hostname
usage: clr system:stuff [--hostname HOSTNAME] [--count COUNT]
                        [hostname] [count]
clr system:stuff: error: argument --hostname: not allowed with argument hostname
```

Reply:
Added a bunch of comments. Also realized vararg handling had a hole it in.